### ISSUE TITLE
Connect: Uncaught Exception on Connection Error

### ISSUE THREAD

**Issue by 3DEsprit** [sanketbajoria/ssh2-promise#13]
**Issue**: Failed connections persist after error, causing the connection to stay open for a long period of time.

**Error**
```
(node:13401) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 192.168.1.101:22
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1162:14)
(node:13401) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block,or by rejecting a promise which was not handled with .catch(). (rejection id: 16)
(node:13401) UnhandledPromiseRejectionWarning: Error: connect ECONNREFUSED 192.168.1.101:22
```

Adding a catch to the wrapper promise causes this error to duplicate (just catching the same error). It seems to stop after 12 occurrences in VSCode, and 30 times in a spec test.

Following the process to see where it's hitting, it looks like this is happening directly on the **emit** method after the **connect** in the _SSHConnection_. After that it throws the unhandled promise, but it doesn't seem to close the connection after that, or even hit the error event. Throwing a catch on the connect does wrap the error, but that's about it. 

<img width="705" alt="sshconnection 182" src="https://user-images.githubusercontent.com/3156278/47330276-3b3cf580-d63d-11e8-9149-d26e9c63f1df.png">

<img width="701" alt="sshconnection 36" src="https://user-images.githubusercontent.com/3156278/47330286-48f27b00-d63d-11e8-8f7a-bd171513cd27.png">

Anyways, I hope that helps, and at least it stops eventually. 😄

**Comment by 3DEsprit** [comment]
Ah-ha, great to know. Thanks!

**Comment by sanketbajoria** [comment]
@3DEsprit ,
Ah, I forgot, i have already defined that in documentation. (Readme)

**Comment by 3DEsprit** [comment]
Thanks!

**Comment by 3DEsprit** [comment]
I completely forgot, is there a way to catch the exceptions thrown? I'm getting this on shell, and spawn even with successful connections, and adding to the connection promise isn't doing it.

**Comment by 3DEsprit** [comment]
It's only the failed connections. I'm just looking to catch the specific connection error to return, instead of the unhandled exception that is thrown. It is not a huge deal, but I thought I would ask.

**Comment by samhon18** [comment]
It seems the promise object only handles the first connection error. Once we enable the reconnect function, the exception thrown by "reconnect" will not call any promise reject callback. And Promise warn it 

The first error throw by connect function was caught and pass to promise reject callback.

> [2018-11-11T00:26:56.949] [ERROR] FilePush - [10.xx.xx.xx] Failed to connect to server { Error: Timed out while waiting for handshake
>     at Timeout._onTimeout (E:\coinSource\DEV\app\node_modules\ssh2\lib\client.js:695:19)
>     at ontimeout (timers.js:436:11)
>     at tryOnTimeout (timers.js:300:5)
>     at listOnTimeout (timers.js:263:5)
>     at Timer.processTimers (timers.js:223:10) level: 'client-timeout' }
> 

The error thrown by the reconnect function are not passed to promise reject callback and we get a unhandledPromiseRejectWarning.

> (node:5228) UnhandledPromiseRejectionWarning: Error: Timed out while waiting for handshake
>     at Timeout._onTimeout (E:\coinSource\DEV\app\node_modules\ssh2\lib\client.js:695:19)
>     at ontimeout (timers.js:436:11)
>     at tryOnTimeout (timers.js:300:5)
>     at listOnTimeout (timers.js:263:5)
>     at Timer.processTimers (timers.js:223:10)
> (node:5228) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
> (node:5228) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> (node:5228) UnhandledPromiseRejectionWarning: Error: Timed out while waiting for handshake
>     at Timeout._onTimeout (E:\coinSource\DEV\app\node_modules\ssh2\lib\client.js:695:19)
>     at ontimeout (timers.js:436:11)
>     at tryOnTimeout (timers.js:300:5)
>     at listOnTimeout (timers.js:263:5)
>     at Timer.processTimers (timers.js:223:10)
> (node:5228) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
> (node:5228) UnhandledPromiseRejectionWarning: Error: Timed out while waiting for handshake
>     at Timeout._onTimeout (E:\coinSource\DEV\app\node_modules\ssh2\lib\client.js:695:19)
>     at ontimeout (timers.js:436:11)
>     at tryOnTimeout (timers.js:300:5)
>     at listOnTimeout (timers.js:263:5)
>     at Timer.processTimers (timers.js:223:10)
> (node:5228) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 3)
> (node:5228) UnhandledPromiseRejectionWarning: Error: Timed out while waiting for handshake
>     at Timeout._onTimeout (E:\coinSource\DEV\app\node_modules\ssh2\lib\client.js:695:19)
>     at ontimeout (timers.js:436:11)
>     at tryOnTimeout (timers.js:300:5)
>     at listOnTimeout (timers.js:263:5)
>     at Timer.processTimers (timers.js:223:10)
> (node:5228) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 4)

**Comment by lqueryvg** [comment]
I'm suffering from this too.

I'm doing `data = await sftp.readdir()` after which I've got some more code I want to execute (if the read works). If the connect fails (e.g. `ECONNREFUSED`), it throws the error, which I catch and handle, but it retries in the background !? It doesn't make any sense. What's the point of throwing an error *and* retrying in the background ? If any of the subsequent retries was to work, the result would be useless to me because the original call stack has gone (because it threw the error).

Surely it should only throw the error once all retry attempts have failed ?

**Comment by sanketbajoria** [comment]
@3DEsprit @lqueryvg 
Didn't get much time to see this issue. 
Will be fixed by next release.

**Comment by sanketbajoria** [comment]
@3DEsprit , @lqueryvg 
Fixed it,

**Comment by 3DEsprit** [comment]
I'll try to find some time this weekend to review, if @lqueryvg doesn't find the time first.

**Comment by lqueryvg** [comment]
@sanketbajoria - yes this looks good to me.

The new behaviour I see is that it it doesn't throw the error until all retries have been exhausted, which is correct IMO.

Good work !

**Comment by sanketbajoria** [comment]
@lqueryvg 
Thank you, for verifying at your end.

Published a new version 0.1.3 to npm.
